emulator/src/hdmi_cec_driver.c

```diff
@@ -0,0 +1,828 @@
+  { DEVICE_POWER_STATUS_OFF, "off" },
+  { DEVICE_POWER_STATUS_STANDBY, "standby" }
+};
+
+static char* GetDeviceTypeStr(device_type_t type)
+{
+  char* result = NULL;
+  for (int i = 0;  i < sizeof (gDIMap) / sizeof (gDIMap[0]);  ++i)
+  {
+    if (type == gDIMap[i].val)
+    {
+        result = gDIMap[i].str_type;
+        break;
+    }
+  }
+  return result;
+}
+
+static device_type_t GetDeviceType(char *type)
+{
+
+  device_type_t result = DEVICE_TYPE_TV;
+
+  for (int i = 0;  i < sizeof (gDIMap) / sizeof (gDIMap[0]);  ++i)
+  {
+    if (!strcmp (type, gDIMap[i].str_type))
+    {
+        result = gDIMap[i].val;   
+        break; 
+    }
+  }
+
+  return result;
+}
+
+static char* GetVendorName(int code)
+{
+  char* result = NULL;
+
+  for (int i = 0;  i < sizeof (gVCMap) / sizeof (gVCMap[0]);  ++i)
+  {
+    if (code == gVCMap[i].code)
+    {
+        result = gVCMap[i].name;   
+        break; 
+    }
+  }
+
+  return result;
+}
+
+static unsigned int GetVendorCode(char *name)
+{
+  unsigned int result = 0;
+
+  for (int i = 0;  i < sizeof (gVCMap) / sizeof (gVCMap[0]);  ++i)
+  {
+    if (!strcmp (name, gVCMap[i].name))
+    {
+        result = gVCMap[i].code;   
+        break; 
+    }
+  }
+
+  return result;
+}
+
+static char* GetDevicePowerStatusStr(device_power_status_t status)
+{
+  char* result = NULL;
+
+  for (int i = 0;  i < sizeof (gPSMap) / sizeof (gPSMap[0]);  ++i)
+  {
+    if (status == gPSMap[i].status)
+    {
+        result = gPSMap[i].str;
+        break; 
+    }
+  }
+  return result;
+}
+
+static device_power_status_t GetDevicePowerStatus(char *name)
+{
+  device_power_status_t result = DEVICE_POWER_STATUS_OFF;
+
+  for (int i = 0;  i < sizeof (gPSMap) / sizeof (gPSMap[0]);  ++i)
+  {
+    if (!strcmp (name, gPSMap[i].str))
+    {
+        result = gPSMap[i].status;
+        break; 
+    }
+  }
+
+  return result;
+}
+
+
+static hdmi_port_type_t GetPortType(char *type)
+{
+  hdmi_port_type_t result = PORT_TYPE_OUTPUT;
+
+  if(!strncmp(type, "in", sizeof("in")))
+  {
+    result = PORT_TYPE_INPUT;
+  }
+  return result;
+}
+
+
+void LoadDevicesInfo (ut_kvp_instance_t* instance, device_info_t* devices, unsigned int nDevices)
+{
+  if (devices != NULL && nDevices > 0)
+  {
+    char *prefix = "hdmicec/devices/";
+    char tmp[256];
+
+    for (int i = 0; i < nDevices; ++i)
+    {
+      char type[16];
+      strcpy(tmp, prefix);
+      int length = snprintf( NULL, 0, "%d", i );
+      snprintf( tmp + strlen(prefix) , length + 1, "%d", i );
+
+      strcpy(tmp + strlen(prefix) + length, "/name");
+      ut_kvp_getStringField(instance, tmp, devices[i].osd_name, MAX_OSD_NAME_LENGTH);
+
+      strcpy(tmp + strlen(prefix) + length, "/active_source");
+      devices[i].active_source = ut_kvp_getBoolField(instance, tmp);
+
+      strcpy(tmp + strlen(prefix) + length, "/pwr_status");
+      ut_kvp_getStringField(instance, tmp, type, sizeof(type));
+      devices[i].power_status = GetDevicePowerStatus(type);
```

Comment:
this function needs to be moved to ut-kvp. Where it's upgraded to support looking up a string and converting it to an unt32_t  or something

 here's an example -> 

```c
`static const char *enumStrings[]=
{
    "XFW_CERTSTORE_STATUS_OK",
    "XFW_CERTSTORE_STATUS_INVALID_INSTANCE",
    "XFW_CERTSTORE_STATUS_INVALID_METHOD",
    "XFW_CERTSTORE_STATUS_IO_FAILURE",
    "XFW_CERTSTORE_STATUS_TRUST_FAILURE",
    "XFW_CERTSTORE_STATUS_INVALID_CERTIFICATE_STORE",
    
    "XFW_CERTSTORE_STATUS_MAX"
};
...
void xfw_instance_attach_status_code_strings( void *pAnyInstance, const char *pStatusCodeStrings[], uint32_t stringTableEntries )
{
    pInstance->pStatusCodeStrings = ( const void **)pStatusCodeStrings;
    pInstance->stringTableEntries = stringTableEntries;

}
...

xfw_instance_get_status_code_string(instance, statusCode)
{
    if ( (uint32_t)statusCode > pInstance->stringTableEntries )
    {
        return invalidString;
    }

    result = pInstance->pStatusCodeStrings[(uint32_t)statusCode];
}


You can extend the table to have two values if you like,, "String", "Value", look up the string return the value.